src/lib.rs

```diff
@@ -0,0 +1,1210 @@
+        if index >= self.len {
+            return None;
+        }
+        let b = self.vec[block_offset::<T>(index)];
+        Some(b & (T::one() << (index % bits_per_block::<T>())) != T::zero())
+    }
+
+    /// Sets the value of the element at position `index` or `None` if out of bounds. Returns
+    /// `true` if this led to a change in the underlying storage or `None` if out of bounds.
+    ///
+    /// # Examples
+    /// ```
+    /// use vob::Vob;
+    /// let mut v = Vob::new();
+    /// v.push(false);
+    /// v.set(0, true);
+    /// assert_eq!(v.get(0), Some(true));
+    /// assert_eq!(v.set(0, false), Some(true));
+    /// assert_eq!(v.set(0, false), Some(false));
+    /// ```
+    pub fn set(&mut self, index: usize, value: bool) -> Option<bool> {
+        if index >= self.len {
+            return None;
+        }
+        let m = T::one() << (index % bits_per_block::<T>());
+        let b = block_offset::<T>(index);
+        let o = self.vec[b];
+        let v = if value {
+                    o | m
+                } else {
+                    o & !m
+                };
+        if o != v {
+            self.vec[b] = v;
+            Some(true)
+        } else {
+            Some(false)
+        }
+    }
+
+    /// Returns an iterator over the slice.
+    ///
+    /// # Examples
+    /// ```
+    /// use vob::Vob;
+    /// let mut v = Vob::new();
+    /// v.push(false);
+    /// v.push(true);
+    /// let mut iterator = v.iter();
+    /// assert_eq!(iterator.next(), Some(false));
+    /// assert_eq!(iterator.next(), Some(true));
+    /// assert_eq!(iterator.next(), None);
+    /// ```
+    pub fn iter(&self) -> Iter<T> {
+        Iter {
+            vob: self,
+            range: 0..self.len
+        }
+    }
+
+
+    /// Returns an iterator which produces the index of each bit set in the Vob.
+    ///
+    /// # Examples
+    /// ```
+    /// use vob::Vob;
+    /// let mut v = Vob::new();
+    /// v.push(false);
+    /// v.push(true);
+    /// let mut iterator = v.iter_set_bits();
+    /// assert_eq!(iterator.next(), Some(1));
+    /// assert_eq!(iterator.next(), None);
+    /// ```
+    pub fn iter_set_bits(&self) -> IterSetBits<T> {
+        IterSetBits {
+            vob: self,
+            range: 0..self.len
+        }
+    }
+
+    /// Returns an iterator which produces the index of each bit set in the Vob.
+    ///
+    /// # Examples
+    /// ```
+    /// use vob::Vob;
+    /// let mut v = Vob::new();
+    /// v.push(false);
+    /// v.push(true);
+    /// let mut iterator = v.iter_unset_bits();
+    /// assert_eq!(iterator.next(), Some(0));
+    /// assert_eq!(iterator.next(), None);
+    /// ```
+    pub fn iter_unset_bits(&self) -> IterUnsetBits<T> {
+        IterUnsetBits {
+            vob: self,
+            range: 0..self.len
+        }
+    }
+
+    /// Return an iterator over the underlying storage blocks. The last block is guaranteed to have
+    /// "unused" bits (i.e. those past `self.len()`) set to 0.
```

Comment:
I see, so perhaps this is why you have the invariant that the internal storage has `0`s at the top.

I wonder if you would get any meaningful speed up by relaxing this constraint (and having, internally, the top bits undefined), and only masking off when the user need to perceive the top bits as `0`s.

Thoughts?

Reply:
I doubt we could measure the difference in any reasonable use case (but I don't know for sure).